### ISSUE TITLE
[reflection]: obj.ReferenceEquals not reflexive

### ISSUE THREAD

**Issue by thautwarm** [fable-compiler/Fable.Python#42]
```F#
type mytype = { a : int ; b : int }

let inline (==) (a: 'a) (b: 'b) = obj.ReferenceEquals(a, b)


let inline get<'a>() = typeof<'a>


let _ =
    let t = get<mytype>()
    printfn "%A" <| (typeof<mytype> == t)
    printfn "%A" <| (t == typeof<mytype>)

(*
False
True
*)
```

This bug is detected for both 0.0.25 and 0.0.26..

**Comment by thautwarm** [comment]
The generated code:

```python
def expr_4() -> TypeInfo:
    return record_type("Prog.mytype", [], mytype, lambda: [["a", int32_type], ["b", int32_type]])


class mytype(Record):
    def __init__(self, a: int, b: int) -> None:
        super().__init__()
        self.a = a or 0
        self.b = b or 0
    

mytype_reflection = expr_4

t : Any = mytype_reflection()

arg10 : bool = mytype_reflection() is t

to_console(printf("%A"))(arg10)

arg10_1 : bool = t == mytype_reflection()

to_console(printf("%A"))(arg10_1)
```

**Comment by thautwarm** [comment]
I think types shall have referential/physical equality, which is now not a truth in Fable.Python.

A simple solution is we design `TypeInfo` as a `reference` type in Python
```python
class CurrentTypeInfo:  # the current impl
      ...

class TypeInfo:   #  has object identity and corresponding __hash__ 
      def __init__(self):
            self.impl = None
```

And any reflection like `expr_4` should return the same object for the same type parameters:

```python
_type_creation_cache_for_expr_4 = {}
def expr_4(*gens) -> TypeInfo:
    if type_info := _type_creation_cache_for_expr_4.get(gens):
        return type_info 
    
    type_info = _type_creation_cache_for_expr_4[gens] = TypeInfo()
    type_info.impl = record_type("Prog.mytype", [], mytype, lambda: [["a", int32_type], ["b", int32_type]])
    return type_info
```

Such solution has several merits that cover the current approach, and promises referential equality for the same instantiation of a generic type constructor:

1. reflection becomes efficient. 
    For instance, automatic serialization needs reflection. I made [a cross-backend Fable alternative to FSharp.Json](https://github.com/thautwarm/FSharp.AJson/blob/main/AJson.fs). However, without efficient reflection, the library is extremely slow, as parsing and to-F#-data conversion themselves are even faster than reflection...
2. recursive references are resolved, just like the current approach
3. correct equality for `TypeInfo` are promised.

**Comment by dbrattli** [comment]
Could we re-use `functool.cache` here?

```py
from typing import Any
from functools import cache
from fable_modules.fable_library.reflection import TypeInfo, int32_type, record_type
from fable_modules.fable_library.string import to_console, printf
from fable_modules.fable_library.types import Record


@cache
def expr_1() -> TypeInfo:
    return record_type("QuickTest.mytype", [], mytype, lambda: [["a", int32_type], ["b", int32_type]])


class mytype(Record):
    def __init__(self, a: int, b: int) -> None:
        super().__init__()
        self.a = a or 0
        self.b = b or 0


mytype_reflection = expr_1

t: Any = mytype_reflection()

arg10: bool = mytype_reflection() is t

to_console(printf("%A"))(arg10)

arg10_1: bool = t == mytype_reflection()

to_console(printf("%A"))(arg10_1)

````

**Comment by thautwarm** [comment]
After pondering for a while, I think simply caching it works... Never mind the above proposal..

**Comment by thautwarm** [comment]
Besides, we should still omit `__eq__`  and `__hash__` for `TypeInfo`, leaving them to use referential equality.

**Comment by thautwarm** [comment]
Oh, `functools.cache` works only for Python >=3.9. 
For Python 3.8 compatibility, We could just implement our own cache. This is safe because the cache entry is always a tuple of TypeInfo, which is a simple case.

```python
def cache(func):
     local_cache = {}
     def instantiate_type(*args):
          if type_info := local_cache.get(args):
                pass
          else:
               type_info = local_cache[args] = func(*args)
          return type_info
     return instantiate_type
```

**Comment by dbrattli** [comment]
Note that `cache` returns the same as `lru_cache(maxsize=None)` so we can just use `lru_cache` instead. I'm making a PR for this right now.

**Comment by thautwarm** [comment]
Thank you!

**Comment by dbrattli** [comment]
I btw had to add a proper `__hash__` method to `TypeInfo` or other tests would start to fail. Fixed in https://www.nuget.org/packages/fable-py/4.0.0-alpha-027

**Comment by thautwarm** [comment]
Thanks! I will do some local tests to figure out why it fails when `__hash__` is missing.

**Comment by thautwarm** [comment]
I got it, I forgot it was a dataclass.

**Comment by thautwarm** [comment]
`@lru_cache` should be `@lru_cache(maxsize=None)`, otherwise the reflection function does not return the same object.

**Comment by dbrattli** [comment]
I don't understand why `maxsize=None` (unbounded) should give different result than `maxsize=128` (default). I can fix it, but the test is working `MyRecord_reflection() is t`. Could you produce a negative test?

```py
@lru_cache
def expr_1035() -> TypeInfo:
    return record_type_1("Fable.Tests.Reflection.MyRecord", [], MyRecord, lambda: [["String", string_type], ["Int", int32_type]])


class MyRecord(Record):
    def __init__(self, String: str, Int: int) -> None:
        super().__init__()
        self.String = String
        self.Int = Int or 0
    

MyRecord_reflection = expr_1035

def test_obj_reference_equals_not_reflexive() -> None:
    t : Any = MyRecord_reflection()
    Testing_equal(True, MyRecord_reflection() is t)
    Testing_equal(True, t == MyRecord_reflection())
```

**Comment by thautwarm** [comment]
The difference can be observed in the following code:

```python
from functools import lru_cache

class S:
    pass

class S:
    def __init__(self, i):
        self.i = i

@lru_cache
def f1(x: int):
    return S(x)

@lru_cache(maxsize=None)
def f2(x: int):
    return S(x)

def test(f):
    res = set()
    for i in range(256 * 100):
        res.add(f(i%256))
    return len(res)

test(f2)  # 25600

test(f1)  # 256
```

LRU cache stores a limited number of the recently used entries.

**Comment by dbrattli** [comment]
But we have only 1 entry. Our function does not take any arguments:

```py
class S:
    def __init__(self, i):
        self.i = i


@lru_cache
def f1():
    return S(42)


@lru_cache(maxsize=None)
def f2():
    return S(42)


def test(f):
    res = set()
    for i in range(256 * 100):
        res.add(f())
    return len(res)


test(f2)  # 1

test(f1)  # 1
```

**Comment by thautwarm** [comment]
IIRC , for generic types, their reflection functions are

```python
def expr_XXX(gen1, gen2, ..., genN) -> TypeInfo:
     ... 
```

**Comment by dbrattli** [comment]
😱 Yes, you are absolutely right. I will fix ...
